src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,464 @@
+    int                      is_abort;
+} ngx_http_lua_worker_thread_ctx_t;
+
+
+static  ngx_http_lua_task_ctx_t   dummy_ctx;
+static  ngx_http_lua_task_ctx_t  *ctxpool = &dummy_ctx;
+static  ngx_uint_t                worker_thread_vm_count;
+
+static ngx_http_lua_task_ctx_t *
+ngx_http_lua_get_task_ctx(lua_State *L, ngx_http_request_t *r)
+{
+    ngx_http_lua_task_ctx_t *ctx = NULL;
+
+    size_t           path_len;
+    const char      *path;
+    size_t           cpath_len;
+    const char      *cpath;
+    lua_State       *vm;
+
+    ngx_http_lua_main_conf_t    *lmcf;
+
+    lmcf = ngx_http_get_module_main_conf(r, ngx_http_lua_module);
+
+    worker_thread_vm_count++;
+
+    if (worker_thread_vm_count > lmcf->worker_thread_vm_pool_size) {
+        return NULL;
+    }
+
+    if (ctxpool->next == NULL) {
+        ctx = calloc(sizeof(ngx_http_lua_task_ctx_t), 1);
+        vm = luaL_newstate();
+        ctx->vm = vm;
+        luaL_openlibs(ctx->vm);
+
+        /* copy package.path and package.cpath */
+        lua_getglobal(L, "package");
+        lua_getfield(L, -1, "path");
+        path = lua_tolstring(L, -1, &path_len);
+        lua_getfield(L, -2, "cpath");
+        cpath = lua_tolstring(L, -1, &cpath_len);
+
+        lua_getglobal(vm, "package");
+        lua_pushlstring(vm, path, path_len);
+        lua_setfield(vm, -2, "path");
+        lua_pushlstring(vm, cpath, cpath_len);
+        lua_setfield(vm, -2, "cpath");
+
+        /* pop path and cpath from L */
+        lua_pop(L, 2);
+
+        /* pop the "package" table */
+        lua_pop(L, 1);
+        lua_pop(vm, 1);
+
+        ngx_thread_mutex_create(&ctx->mutex, ngx_cycle->log);
+
+    } else {
+        ctx = ctxpool->next;
+        ctxpool->next = ctxpool->next->next;
+    }
+
+    return ctx;
+}
+
+
+static void
+ngx_http_lua_free_task_ctx(ngx_http_lua_task_ctx_t *ctx)
+{
+    ctx->next = ctxpool->next;
+    ctxpool->next = ctx;
+    /*  clean Lua stack */
+    lua_settop(ctx->vm, 0);
+
+    worker_thread_vm_count--;
+}
+
+
+static int
+ngx_http_lua_xcopy(lua_State *from, lua_State *to, int idx,
+                         const int allow_nil)
+{
+    size_t           len = 0;
+    const char      *str;
+
+    switch (lua_type(from, idx)) {
+    case LUA_TBOOLEAN:
+        lua_pushboolean(to, lua_toboolean(from, idx));
+        return LUA_TBOOLEAN;
+
+    case LUA_TLIGHTUSERDATA:
+        lua_pushlightuserdata(to, lua_touserdata(from, idx));
+        return LUA_TLIGHTUSERDATA;
+
+    case LUA_TNUMBER:
+        lua_pushnumber(to, lua_tonumber(from, idx));
+        return LUA_TNUMBER;
+
+    case LUA_TSTRING:
+        str = lua_tolstring(from, idx, &len);
+        lua_pushlstring(to, str, len);
+        return LUA_TSTRING;
+
+    case LUA_TTABLE:
+        lua_newtable(to);
+        /* to positive number */
+        if (idx < 0) {
+            idx = lua_gettop(from) + idx + 1;
+        }
+
+        lua_pushnil(from);
+
+        while (lua_next(from, idx)) {
```

Comment:
```suggestion
        while (lua_next(from, idx) != 0) {
```
this is the recommended style.